### PR TITLE
Classification and Clustering Benchmarking

### PR DESCRIPTION
**Pull request by WiscEvan**
## Overview

1. :art: Add `autometa-benchmark` entrypoint under `autometa.validation.benchmark` module
2. :art::green_apple: Add `BENCHMARK_CLUSTERING` nextflow process
3. :art::green_apple: Add `BENCHMARK_CLASSIFICATION` nextflow process
4. :art::green_apple: Add `publishDir` operator to `BINNING_SUMMARY` process.

### `autometa-benchmark` entrypoint

Add `autometa-benchmark` entrypoint for classification and clustering benchmarking against autometa reference communities. This adds functionality to easily compute classification and clustering benchmarking statistics of test datasets against the current 'gold' standard (Set as Autometa v.1 predictions).

### `BENCHMARK_CLUSTERING` nextflow process

~**NOTE: PR #138 entrypoint required for Nextflow process**~

This process takes as input all binning/clustering predictions as well as the `community` value associated with these binning predictions. ~This requires the to-be-added `autometa-datasets` entrypoint from PR #138.~ Two tables are written (wide and long format) with clustering metrics computed using scikit-learn's clustering metrics.

### Example Figure: Simulated Community Clustering Metrics

These metrics were plotted using the long-formatted (`--output-long`) output file generated by `autometa-benchmark` with `--benchmark=clustering`

simulated_community_clustering_evaluation_metrics

### `BENCHMARK_CLASSIFICATION` nextflow process

~**NOTE: PR #138 entrypoint required for Nextflow process**~

This process takes as input all taxonomic predictions (tab-delimited file of contig and taxid columns) as well as the `community` value associated with these taxonomic predictions. ~This requires the to-be-added `autometa-datasets` entrypoint from PR #138.~ One table is written with overall classification metrics computed using scikit-learn's classification metrics. Additionally a classification report is written per community. Metrics are computed using taxids such that all taxonomic predictions may be classified (not only canonical-ranks). This allows assessment of taxonomic predictions grouped by canonical and non-canonical ranks.

### Example Figure: Median precision vs. recall of taxonomic predictions grouped by non-canonical ranks. 

These metrics were plotted using the output file generated by `autometa-benchmark` with `--benchmark=classification`

**Figures are faceted by simulated community dataset**

support values are the number of taxa within each taxonomic label class

median_scores_by_rank

### CODE REVIEW

**Comment by WiscEvan** [comment]
# Benchmarking datasets/commands

## Testing Benchmarking Command

Written in the new `nextflow/benchmarking-tasks.nf` script process [BENCHMARK_CLUSTERING](https://github.com/KwanLab/Autometa/pull/141/files?file-filters%5B%5D=.nf&file-filters%5B%5D=.py&file-filters%5B%5D=No+extension#diff-f313c8ca2757c38026de47359b8b7dd02e76bb0754a1edc0b4b65c2045723e7bR10-R37), two commands are performed. The first is to retrieve the respective test dataset. In the example above multiple simulated communities were separately downloaded and compared until their results were finally collected and visualized as follows:

1. [Retrieve the required test data file](#retrieve-the-required-test-data).
2. Perform benchmarking using the input test data and respective results (classification or binning)

### Retrieve the required test data

Awaiting PR #138 for the `autometa-datasets` entrypoint. Until then you will need to manually download the respective communities reference assignments file as well as the 'gold-standard' assignment (binning/taxonomy for clustering/classification comparison, respectively) uploaded from autometa version 1.0

**NOTE: This entrypoint functionality requires PR#138**

```bash

# Note: community is the test dataset that was used for clustering or classification. e.g.
# choices: 78, 156, 312, 625, 1250, 2500, 5000, 10000, etc.
communities = (78, 156, 312, 625, 1250, 2500, 5000, 10000)
for community in ${communities[@]};do
    autometa-datasets \
      --community ${community} \
      --file reference_assignments,binning \
      --output ${community}.reference.tsv.gz,${community}.binning.tsv.gz

    autometa-benchmark \
        --benchmark clustering \
        --predictions ${community}.binning.tsv.gz \
        --reference ${community}.reference.tsv.gz \
        --output-wide ${community}.clustering_benchmarks.wide.tsv.gz \
        --output-long ${community}.clustering_benchmarks.long.tsv.gz
done
# NOTE: Concatenate the long output for figure 1 visualization (concatenation is in code block below)
```

### Aggregate across simulated communities

```python
import pandas as pd
import glob
df = pd.concat([
    pd.read_csv(fp, sep="\t", index_col="dataset")
    for fp in glob.glob("*.clustering_benchmarks.long.tsv.gz")
])
df.to_csv("benchmarks.tsv", sep='\t', index=False, header=True)
```


## Code for Visualization of Simulated Community Benchmarking Results

### Figure 1: Simulated Communities Clustering Evaluation Metrics

Note: This benchmarks file was computed using the reference assignments and gold standards across the simulated community datasets. This requires downloading the aforementioned reference_assignments and binning files and aggregating these into a single file for visualization in R. 

```R
# Figure 1: Simulated Communities Clustering Evaluation Metrics
library(ggplot2)
library(dplyr)
setwd("~/benchmarking/param_sweep/v1.0_standards/") # directory containing results
# metrics in long format (autometa-benchmark --output-long 'benchmarks.tsv')
df <- read.table("benchmarks.tsv", sep='\t', header = T)
df$dataset <- factor(df$dataset, levels = c("78.125Mbp",
                                            "156.25Mbp",
                                            "312.5Mbp",
                                            "625Mbp",
                                            "1250Mbp",
                                            "2500Mbp",
                                            "5000Mbp",
                                            "10000Mbp"))
ggplot(data=df, 
                mapping=aes(x=reorder(dataset, desc(dataset)), 
                            y=score, fill=dataset)) +
  geom_bar(stat='identity', position='dodge') + 
  facet_wrap(~ metric) +
  theme_bw() +
  coord_flip() +
  labs(
    x="Simulated Community",
    y='Clustering Metric Score', 
    title="Simulated Communities Clustering Evaluation Metrics")
```

## Figure 2: K-mer parameter sweep clustering evaluation

This logic will follow a similar behavior to above, except we are also providing our binning predictions with the `autometa-benchmark` entrypoint. We will use the aggregated results here to assess benchmarks across the k-mer parameter sweep.

### Benchmark 'gold-standard' test data and benchmarking experiments against reference assignments

```bash
binning="Path to your autometa binning predictions file"
# Now benchmark inputs and previous gold-standard against community reference assignments
community=78
autometa-benchmark \
    --benchmark clustering \
    --predictions $binning ${community}.binning.tsv.gz \
    --reference ${community}.reference.tsv.gz \
    --output-wide ${community}.clustering_benchmarks.wide.tsv.gz \
    --output-long ${community}.clustering_benchmarks.long.tsv.gz
```

### Aggregate all communities clustering benchmarks

```python
import pandas as pd
import glob
long_df = pd.concat([
    pd.read_csv(fp, sep="\t", index_col="dataset")
    for fp in glob.glob("*.clustering_benchmarks.long.tsv.gz")
])

wide_df = pd.concat([
    pd.read_csv(fp, sep="\t", index_col="dataset")
    for fp in glob.glob("*.clustering_benchmarks.wide.tsv.gz")
])

# Now write to main files for visualization
long_df.to_csv("clustering_benchmarks.long.tsv")
wide_df.to_csv("clustering_benchmarks.wide.tsv")
```

### Retrieve parameters from filename and add to dataframe

```python
import pandas as pd

df = pd.read_csv(
    "clustering_benchmarks.long.tsv",
    sep='\t',
    index_col='dataset'
)

# NOTE: Reference cols follow:
# 625Mbp_reference_binning
# NOTE: index follows format written to output file paths from parameter sweep:
# [community].[kingdom].pca[pca_dims].[embed_method]_[embed_dims].[cluster_method].binning.tsv.gz
# e.g.
# 156Mbp.bacteria.pca350.bhsne_10.hdbscan.binning.tsv.gz
# 'gold-standard' is named 'reference_assignments.tsv.gz'
# We know the gold-standard parameters so these are placed 
# in the column if the dataset is the reference.

# Retrieve cluster method and assign to 'cluster_method' column
df['cluster_method'] = df.index.map(
    lambda i: 
    'dbscan' if "reference" in i
     else i.split("pca")[-1].split(".")[2]
)

# Retrieve pca dimensions and assign to 'pca_dims' column
df['pca_dims'] = df.index.map(
  lambda i:
  '50' if "reference" in i
  else i.split("pca")[-1].split(".")[0]
)

# Retrieve embedding dimensions and assign to 'embed_dims' column
df['embed_dims'] = df.index.map(
  lambda i:
  '2' if "reference" in i
  else i.split("pca")[-1].split(".")[1].split("_")[1]
)

# Retrieve embedding method and assign to 'embed_dims' column
df['embed_method'] = df.index.map(
  lambda i:
  "bhsne" if "reference" in i
   else i.split("pca")[-1].split(".")[1].split("_")[0]
)

# Some simulated community names have more significant digits than their respective folders
mappings = {'156Mbp':'156.25Mbp', '312Mbp':'312.5Mbp', '78Mbp':'78.125Mbp'}

df['community'] = df.index.map(lambda i: i.split("_")[0] if "reference" in i else i.split(".bacteria")[0])

# Write out new table to plot the second figure
df.to_csv("clustering_benchmarks.long.tsv", sep='\t', index=True, header=True)
```

### Plot k-mer parameter sweep clustering evaluation

```R
library(ggplot2)
library(dplyr)

wide_df <- read.table("clustering_benchmarks.wide.tsv", sep='\t', header = TRUE)
colnames(wide_df)

p <- ggplot(data=wide_df, mapping=aes(x=embed_dims,  y=adjusted.rand.score)) +
  geom_point(aes(color=cluster_method, shape=embed_method)) + 
  theme_bw() +
  facet_wrap(~ community) +
  labs(
    x="Embedding dimensions",
    y='Adjusted Rand Score', 
    title="Simulated Community k-mer parameter sweep clustering evaluation benchmarking")

ggsave("simulated_community_kmer_param_sweep_ARI_benchmarking.pdf", p)
```

### k-mer parameter sweep homogeneity and completeness benchmarking

```R
q <- ggplot(data=wide_df, mapping=aes(x=homogeneity.score, y=completeness.score)) +
  geom_point(aes(color=cluster_method, shape=embed_method)) + 
  theme_bw() +
  facet_wrap(~community) +
  labs(
    x="Homogeneity Score",
    y='Completeness Score', 
    title="Simulated Community k-mer parameter sweep clustering homogeneity and completeness evaluation")
q
ggsave("simulated_community_kmer_param_sweep_homogeneity_and_completeness_benchmarking.pdf", q)
```

# Taxon-profiling benchmarking

### Perform taxonomic classification benchmarking with `autometa-benchmark` across all simulated communities

```bash
autometa-benchmark \
    --benchmark classification \
    --predictions $taxonomy ${community}.taxonomy.tsv.gz \
    --reference ${community}.reference.tsv.gz \
    --output-wide ${community}.classification_benchmarks.wide.tsv.gz \
    --output-classification-reports taxa_reports \
    --ncbi /ncbi
```

## 

### Subset main output file into the two separate averages that are written per dataset 

```python
# Concatenate all classification benchmarks across datasets
df = pd.concat([
  pd.read_csv(fp, sep='\t', )
  for fp in glob.glob("*.classification_benchmarks.tsv.gz")
])
# Write out for weighted average benchmark
weighted_avg_df = df[df.average == 'weighted avg']
weighted_avg_df.to_csv(
    "simulated_classification_benchmarks.weighted_avg.tsv",
    sep='\t',
    index=False,
    header=True
)
# Write out for samples average benchmark
samples_avg_df = df[df.average == 'samples avg']
samples_avg_df.to_csv(
    "simulated_classification_benchmarks.samples_avg.tsv",
    sep='\t',
    index=False,
    header=True
)
```

### Taxon classification samples average of precision and recall

```R
library(ggplot2)
# install.packages("ggforce")
library(ggforce)
library(dplyr)

df <- read.table("simulated_classification_benchmarks.samples_avg.tsv", header = T)

colnames(df)
df
df$precision <- as.numeric(as.vector(df$precision))
df$recall <- as.numeric(as.vector(df$recall))

p<- arrange(df, precision, recall) %>% ggplot( 
  mapping=aes(
    x=precision,
    y=recall)) + 
  geom_point(aes(color=dataset, size=support)) +
  # ggforce::geom_mark_ellipse(aes(label = norm_method, group = norm_method)) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
  labs(
    x="Precision",
    y="Recall", 
    title="Samples average of precision and recall for taxon classification across simulated communities"
  )
p
ggsave("samples_average_precision_recall_taxon_classification.pdf", p)
```

### Taxon classification weighted average of precision and recall

```R
df <- read.table("simulated_classification_benchmarks.weighted_avg.tsv", header = T)
colnames(df)
df
df$precision <- as.numeric(as.vector(df$precision))
df$recall <- as.numeric(as.vector(df$recall))

q <- arrange(df, precision, recall) %>% ggplot( 
  mapping=aes(
    x=precision,
    y=recall)) + 
  geom_point(aes(color=dataset, size=support)) +
  # ggforce::geom_mark_ellipse(aes(label = norm_method, group = norm_method)) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
  labs(
    x="Precision",
    y="Recall", 
    title="Weighted average of precision and recall for taxon classification across simulated communities"
  )
q
ggsave("weighted_average_precision_recall_taxon_classification.pdf", q)
```

## Median precision and recall scores across ranks

### Aggregate taxa classification reports across simulated communities to compute median scores

```python
import os
import pandas as pd
import glob

reports = {os.path.basename(fp).split(".taxa_")[0]:fp for fp in glob.glob("taxa_reports/*_classification_report.tsv.gz")}
metrics = []
for dataset,fpath in reports.items():
    df = pd.read_csv(fpath, sep='\t')
    df = df[df.support != 0.0]
    for rank,dff in df.groupby("rank"):
        medians = dff.median()
        metric = {
            "community": dataset,
            "rank": rank,
            "median_precision":medians.precision,
            "median_recall": medians.recall,
            "median_f1_score": medians['f1-score'],
            "total_support": dff.support.sum(),
        }
        metrics.append(metric)
    main_df = pd.DataFrame(metrics)

main_df.to_csv("taxa_classification_reports_median_scores.tsv", sep='\t', index=False, header=True)
```

### Now visualize in R

```R
df <- read.table("simulated_communities_taxa_classifications_by_rank.tsv", header = T)

colnames(df)
df
df$median_precision <- as.numeric(as.vector(df$median_precision))
df$median_recall <- as.numeric(as.vector(df$median_recall))

g <- arrange(df, median_precision, median_recall) %>% ggplot( 
  mapping=aes(
    x=median_precision,
    y=median_recall)) + 
  geom_point(aes(color=rank, size=total_support)) +
  # ggforce::geom_mark_ellipse(aes(label = norm_method, group = norm_method)) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
  facet_wrap(~community) +
  labs(
    x="Median Precision",
    y="Median Recall", 
    title="Medians of precision and recall for taxon classification across ranks faceted by simulated communities"
  )
g
ggsave("median_scores_by_rank.pdf", g)
```

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -1,11 +1,10 @@
-exclude: "^autometa/validation"
 repos:
   - repo: https://github.com/psf/black
-    rev: stable
+    rev: 21.5b1
```
[the remark this answers is not in the document]
Comment (WiscEvan) [reply .pre-commit-config.yaml:3]: @KwanLab/autometa

**Comment by jason-c-kwan** [comment]
BTW in bash and zsh the following valid syntax:

```
communities = (78, 156, 312, 625, 1250, 2500, 5000, 10000)
```

It won't accept spaces either side of the `=`, so the following works:

```
communities=(78, 156, 312, 625, 1250, 2500, 5000, 10000)
```

File: autometa/validation/benchmark.py [hunk 2]
```diff
@@ -0,0 +1,533 @@
...
+        help="Path to community reference assignments",
+        metavar="filepath",
+        required=True,
+    )
+    parser.add_argument(
+        "--output-wide",
+        help="Path to write benchmarking evaluation metrics (each metric receives its own column) (Default: `benchmark_type`_benchmarks.tsv.gz",
+        metavar="filepath",
+        required=False,
+    )
```
[the remark this answers is not in the document]
Comment (WiscEvan) [reply autometa/validation/benchmark.py:475]: The filenames are being updated in `main(...)`

e.g.

```python
output_wide = (
        f"{args.benchmark}_benchmarks.tsv.gz"
        if not args.output_wide
        else args.output_wide
    )
```

**Comment by jason-c-kwan** [comment]
I updated the comment above because the following gave me an error:

```
--file-names reference_assignments.tsv.gz,binning.tsv.gz,taxonomy.tsv.gz
```

Replacing the commas with spaces works.

File: autometa/validation/benchmark.py [hunk 3]
```diff
@@ -0,0 +1,638 @@
...
+        "--benchmark",
+        help="Type of benchmarking to perform",
+        choices={"clustering", "classification", "binning-classification"},
+        required=True,
+    )
+    parser.add_argument(
+        "--predictions",
+        help="Path to Autometa predictions (May specify multiple if they all correspond to the same `--reference` community ",
+        metavar="filepath",
+        nargs="*",
```
Comment (WiscEvan) [inline autometa/validation/benchmark.py:558]:
```suggestion
        nargs="+",
```

File: autometa/validation/benchmark.py [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+    )
```
Comment (WiscEvan) [inline autometa/validation/benchmark.py:559]:
```suggestion
    parser.add_argument(
        "--predictions",
        help="Path to Autometa predictions (May specify multiple if they all correspond to the same `--reference` community ",
        metavar="filepath",
        nargs="+",
    )
```

File: autometa/validation/benchmark.py (hunk 3 above)
Comment (WiscEvan) [inline autometa/validation/benchmark.py:558]:
```suggestion
        nargs="+",
```

**Comment by Sidduppal** [comment]
You should add the steps to benchmaark in the documentation. I think adding [this](https://github.com/KwanLab/Autometa/pull/141#issuecomment-847348201) comment will be really helpful for anyone trying to reproduce our results.

**Comment by chasemc** [comment]
> You should add the steps to benchmaark in the documentation. I think adding [this](https://github.com/KwanLab/Autometa/pull/141#issuecomment-847348201) comment will be really helpful for anyone trying to reproduce our results.
> Also, adding the description of each of the benchmarking metrics will help in interpreting the results as well.

I've put all of that into an Rmarkdown document that people could run as well and get a report back. Haven't committed or posted that here- waiting on @WiscEvan to send me server paths to binning outputs so I don't have to bin all the datasets again.

**Review by Sidduppal — APPROVED** [review]
Ran the entry points with all three flags. 
Everything seems to be working well 👍

**Pull request merged**